clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc

```diff
@@ -4771,6 +4962,122 @@ sycl::event engine_ext::async_rnn_backward(
   return e;
 }
 // DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|get_dropout_state_size|dpct::dnnl
+// DPCT_DEPENDENCY_BEGIN
+// DnnlUtils|get_dropout_state_size_decl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+inline
+size_t engine_ext::get_dropout_state_size(){
+  return _random_engine_state_size;
+}
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|get_dropout_workspace_size|dpct::dnnl
+// DPCT_DEPENDENCY_BEGIN
+// DnnlUtils|get_dropout_workspace_size_decl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+inline size_t
+engine_ext::get_dropout_workspace_size(const memory_desc_ext &src_desc) {
+  return src_desc.get_size();
+}
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|async_dropout_forward|dpct::dnnl
+// DPCT_DEPENDENCY_BEGIN
+// DnnlUtils|async_dropout_forward_decl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+inline
+sycl::event engine_ext::async_dropout_forward(dropout_desc &desc,
+                                              const memory_desc_ext &src_desc,
+                                              void *src,
+                                              const memory_desc_ext &dst_desc,
+                                              void *dst, void *workspace,
+                                              size_t workspace_size) {
+  if (workspace_size < src_desc.get_size()) {
+    throw std::runtime_error("async_dropout_forward: no sufficient workspace.");
+  }
+  float p = desc.get_probability();
+  if (p == 1.f) {
+    return _q->memset(dst, 0, dst_desc.get_size());
+  } else if (p == 0.f) {
+    return async_reorder(1.f, src_desc, src, 0.f, dst_desc, dst);
+  }
+
+  float scale_factor = 1.f / (1.f - p);
+  void *cache = workspace;
+
+  memory_desc_ext rng_data_desc(
+      ::dnnl::memory::desc(src_desc.get_dims(), ::dnnl::memory::data_type::s32,
+                           src_desc.get_strides()));
+  if (src_desc.get_desc().get_data_type() != ::dnnl::memory::data_type::s32) {
+    cache = allocate(rng_data_desc);
+  }
+
+  desc.generate(_q, _random_engine_state_size, rng_data_desc.get_element_num(),
+                (std::int32_t *)cache);
+
+  if (cache == workspace) {
+    async_scale(scale_factor, src_desc, workspace);
+  } else {
+    async_reorder(scale_factor, rng_data_desc, cache, 0.f, src_desc, workspace);
+  }
+
+  auto execution_args = new std::unordered_map<int, ::dnnl::memory>{
+      {DNNL_ARG_SRC_0, ::dnnl::memory(src_desc.get_desc(), _eng, src)},
+      {DNNL_ARG_SRC_1, ::dnnl::memory(src_desc.get_desc(), _eng, workspace)},
+      {DNNL_ARG_DST, ::dnnl::memory(dst_desc.get_desc(), _eng, dst)}};
+
+  auto primitive = create_forward_primitive<::dnnl::binary>(
+      ::dnnl::algorithm::binary_mul, src_desc.get_desc(), src_desc.get_desc(),
+      dst_desc.get_desc());
+
+  auto e = execute_primitive(primitive, execution_args, {});
+
+  if (cache != workspace) {
+    _q->submit([&](sycl::handler &cgh) {
+      cgh.depends_on(e);
+      cgh.host_task([=] { sycl::free(cache, *_q); });
+    });
+  }
```

Comment:
I am a bit confused: do you assume that the queue is in-order or not?
Asking because between desc.generate (L5020) and execute_primitive (L5038), you did not have any sync on event, whereas you have one between execute_primitive (L5038) and the host_task to free the cache (L5042)

Reply:
Yes, In order to align with CUDA, we use in-order queue.